internal/config/cacher.go

```diff
@@ -0,0 +1,203 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"sync"
+	"time"
+)
+
+// Cacher will detect which config values are read when constructing a value.
+// Then, when further requests are made, it will return the same value as long
+// as the config values which were used don't change.
+type Cacher struct {
+	cfg View
+	m   sync.Mutex
+
+	r *rememberingView
+	v interface{}
+}
+
+// NewCacher returns a cacher which uses cfg to detect relevant changes.
+func NewCacher(cfg View) *Cacher {
+	return &Cacher{
+		cfg: cfg,
+	}
+}
+
+// Get will call f the first time it is called.  It remembers which values are
+// read from the provided View, and the value returned.  When Get is next
+// called, it will check if any of the used values have changed.  If they
+// haven't, it will return the cache'd value.  Otherwise it will call f again.
+// If f returns an error, the value will not be cached or returned.
+func (c *Cacher) Get(f func(cfg View) (interface{}, error)) (interface{}, error) {
```

Comment:
Can we declare a type for this function and document what you expect this function written by the caller to do?

Reply:
How's this for documentation. (On Cacher and on *cacher.Get)